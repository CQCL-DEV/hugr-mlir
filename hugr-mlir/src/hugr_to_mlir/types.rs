use crate::mlir::hugr::ffi::mlirHugrTranslateStringRefToMLIRFunction;
use crate::{mlir, Error, Result};
use hugr::hugr::IdentList;
use melior::Context;
use std::borrow::Borrow;
use std::ops::Deref;

pub fn extension_id_to_extension_attr<'c>(
    context: &'c Context,
    id: &'_ hugr::extension::ExtensionId,
) -> mlir::hugr::ExtensionAttr<'c> {
    crate::mlir::hugr::ExtensionAttr::new(context, id.borrow())
}

pub fn extension_set_to_extension_set_attr<'c>(
    context: &'c Context,
    set: &'_ hugr::extension::ExtensionSet,
) -> mlir::hugr::ExtensionSetAttr<'c> {
    crate::mlir::hugr::ExtensionSetAttr::new(
        context,
        set.iter()
            .map(|x| extension_id_to_extension_attr(context, x))
            .collect::<Vec<_>>(),
    )
}

pub fn hugr_to_mlir_type<'c>(
    ctx: &'c Context,
    type_: &hugr::types::Type,
) -> Result<melior::ir::Type<'c>, Error> {
    use hugr::types::TypeEnum;
    match type_.as_type_enum() {
        TypeEnum::Sum(sum_type) => {
            let mut alts = std::vec::Vec::<_>::new();
            for i in 0usize.. {
                if let Some(t) = sum_type.get_variant(i) {
                    alts.push(hugr_to_mlir_type(ctx, t)?)
                } else {
                    break;
                }
            }
            Ok(mlir::hugr::SumType::new(ctx, alts).into())
        }
        TypeEnum::Tuple(row) => Ok(melior::ir::r#type::TupleType::new(
            ctx,
            collect_type_row::<Vec<_>>(ctx, row)?.as_slice(),
        )
        .into()),
        TypeEnum::Extension(ref custom_type) => {
            let mb_t = match custom_type.extension().deref() {
                "prelude" => match custom_type.name().as_str() {
                    "usize" => Some(unsafe {
                        melior::ir::Type::from_raw(mlir_sys::mlirIndexTypeGet(ctx.to_raw()))
                    }),
                    _ => None
                },
                "arithmetic.int.types" => match custom_type.name().as_str() {
                    "int" => {
                        assert!(!custom_type.args().is_empty());
                        if let hugr::types::type_param::TypeArg::BoundedNat { n } =
                            custom_type.args()[0]
                        {
                            Some(melior::ir::r#type::IntegerType::new(ctx, n as u32).into())
                        } else {
                            panic!(
                                "int type does not have a bounded nat arg: {:?}",
                                custom_type
                            )
                        }
                    }
                    _ => None
                },
                "arithmetic.float.types" => match custom_type.name().as_str() {
                    "float64" => {
                        assert!(custom_type.args().is_empty());
                        Some(unsafe {
                            melior::ir::Type::from_raw(mlir_sys::mlirF64TypeGet(ctx.to_raw()))
                        })
                    }
                    "float32" => {
                        assert!(custom_type.args().is_empty());
                        Some(unsafe {
                            melior::ir::Type::from_raw(mlir_sys::mlirF32TypeGet(ctx.to_raw()))
                        })
                    }
                    _ => None
                },
                _ => None
            };
            if let Some(t) = mb_t {
                Ok(t)
            } else {
                Ok(crate::mlir::hugr::OpaqueType::new(
                    custom_type.name().as_ref(),
                    extension_id_to_extension_attr(ctx, custom_type.extension()),
                    collect_type_args::<Vec<_>>(ctx, custom_type.args())?,
                    mlir::hugr::TypeConstraintAttr::new(ctx, custom_type.bound()),
                    ).into())
            }
        },
        TypeEnum::Alias(ref alias_decl) => Ok(mlir::hugr::AliasRefType::new(
            mlir::hugr::ExtensionSetAttr::new(ctx, []),
            melior::ir::attribute::FlatSymbolRefAttribute::new(ctx, alias_decl.name.as_ref())
                .into(),
            mlir::hugr::TypeConstraintAttr::new(ctx, alias_decl.bound),
        )
        .into()),
        TypeEnum::Function(ref function_type) if function_type.params().is_empty() => {
            Ok(hugr_to_mlir_function_type(ctx, function_type.body())?.into())
        }
        TypeEnum::Function(ref _function_type) => {
            panic!("unimplemented: TypeEnum::Function with params")
        }
        TypeEnum::Variable(_size, _bound) => panic!("unimplemented: TypeEnum::Variable"),
    }
}

pub fn hugr_to_mlir_function_type<'a>(
    ctx: &'a Context,
    type_: &'_ hugr::types::FunctionType,
) -> Result<mlir::hugr::HugrFunctionType<'a>, Error> {
    let type_extensions = extension_set_to_extension_set_attr(ctx, &type_.extension_reqs);
    let type_inputs = collect_type_row::<Vec<_>>(ctx, &type_.input)?;
    let type_outputs = collect_type_row::<Vec<_>>(ctx, &type_.output)?;
    Ok(mlir::hugr::HugrFunctionType::new(
        type_extensions,
        melior::ir::r#type::FunctionType::new(ctx, type_inputs.as_slice(), type_outputs.as_slice()),
    ))
}

pub fn collect_type_row_vec<'c>(
    context: &'c Context,
    row: &'_ hugr::types::TypeRow,
) -> Result<Vec<melior::ir::Type<'c>>> {
    collect_type_row(context, row)
}

pub fn collect_type_row<'c, R: FromIterator<melior::ir::Type<'c>>>(
    context: &'c Context,
    row: &'_ hugr::types::TypeRow,
) -> Result<R, Error> {
    row.iter().map(|x| hugr_to_mlir_type(context, x)).collect()
}

pub fn collect_type_args<'c, 'a, R: FromIterator<melior::ir::Attribute<'c>>>(
    ctx: &'c Context,
    args: impl IntoIterator<Item = &'a hugr::types::TypeArg>,
) -> Result<R, Error> {
    use hugr::types::TypeArg;
    args.into_iter()
        .map(|x| match x {
            TypeArg::Type { ty } => {
                Ok(melior::ir::attribute::TypeAttribute::new(hugr_to_mlir_type(ctx, ty)?).into())
            }
            &TypeArg::BoundedNat { n } => Ok(melior::ir::attribute::IntegerAttribute::new(
                n as i64,
                melior::ir::r#type::IntegerType::new(ctx, 64).into(),
            )
            .into()),
            &TypeArg::Opaque { .. } => todo!(),
            TypeArg::Sequence { ref elems } => Ok(melior::ir::attribute::ArrayAttribute::new(
                ctx,
                &collect_type_args::<Vec<_>>(ctx, elems)?,
            )
            .into()),
            TypeArg::Extensions { es } => Ok(extension_set_to_extension_set_attr(ctx, es).into()),
            ta => panic!("unimplemented type_arg: {:?}", ta),
        })
        .collect::<Result<R, Error>>()
}
