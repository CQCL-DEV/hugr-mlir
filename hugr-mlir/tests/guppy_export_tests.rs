use rstest::{fixture, rstest};
use std::fs;

use hugr_mlir::Result;

#[fixture]
pub fn test_context() -> melior::Context {
    let dr = melior::dialect::DialectRegistry::new();
    let hugr_dh = hugr_mlir::mlir::hugr::get_hugr_dialect_handle();
    hugr_dh.insert_dialect(&dr);
    let ctx = melior::Context::new();
    ctx.append_dialect_registry(&dr);
    ctx.load_all_available_dialects();
    ctx
}

#[rstest]
fn test_guppy_exports(test_context: melior::Context) -> Result<()> {
    use hugr_mlir::hugr_to_mlir::hugr_to_mlir;
    // unsafe { mlir_sys::mlirEnableGlobalDebug(true); }
    insta::glob!("guppy-exports", "*.json", |path| {
        let bytes = fs::read(path).unwrap();
        let ul = melior::ir::Location::new(&test_context, path.to_str().unwrap(), 0, 0);
        let hugr = serde_json::from_slice::<hugr::Hugr>(&bytes).unwrap();
        let mlir_mod = hugr_to_mlir(ul, &hugr).unwrap();
        // println!("{:?}:{}", path, mlir_mod.as_operation());
        assert!(mlir_mod.as_operation().verify());
        insta::assert_display_snapshot!(mlir_mod.as_operation());

        // let hugr2 = mlir_to_hugr(&mlir_mod.as_operation()).unwrap();
        // let mlir_mod2 = hugr_to_mlir(ul, &hugr2).unwrap();
        // assert_eq!(mlir_mod2.as_operation(), mlir_mod.as_operation());
    });
    Ok(())
}
