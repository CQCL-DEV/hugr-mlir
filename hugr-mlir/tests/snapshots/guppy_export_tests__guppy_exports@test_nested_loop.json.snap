---
source: hugr-mlir/tests/guppy_export_tests.rs
expression: mlir_mod.as_operation()
input_file: hugr-mlir/tests/guppy-exports/test_nested_loop.json
---
module {
  hugr.module {
    func @foo[](%arg0: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg1: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> {
      %0 = dfg input extensions [] %arg0, %arg1 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> {
      ^bb0(%arg2: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg3: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
        %1 = cfg %arg2, %arg3 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> {
        ^bb0(%arg4: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg5: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
          %2:4 = dfg input extensions [] %arg4, %arg5 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg7: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            const @const_25 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 0 : i6
            %29 = load_constant @const_25 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %30 = make_tuple()
            %31 = tag 0 %30 : tuple<> -> <tuple<>>
            output %31, %29, %arg6, %arg7 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %2#0, %2#1, %2#2, %2#3 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb2
        ^bb1(%3: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb4
          output %3 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
        ^bb2(%4: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %5: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // 2 preds: ^bb0, ^bb7
          %7 = dfg input extensions [] %4, %5, %6 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>> {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg7: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg8: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            const @const_30 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 0 : i6
            %29 = load_constant @const_30 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %30 = ext_op []"igt_s" %arg7, %29 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr.sum<tuple<>, tuple<>>
            %31 = make_tuple(%arg6, %arg7, %arg8 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>)
            %32 = make_tuple(%arg6 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>)
            %33 = conditional(%30, %31, %32 : !hugr.sum<tuple<>, tuple<>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>> {
            ^bb0(%arg9: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg10: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>):
              %34 = dfg input extensions [] %arg9, %arg10 : (tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>> {
              ^bb0(%arg11: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg12: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>):
                %35 = tag 0 %arg11 : tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>> -> <tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>>
                output %35 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>>
              }
              output %34 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>>
            }, {
            ^bb0(%arg9: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg10: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>):
              %34 = dfg input extensions [] %arg9, %arg10 : (tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>> {
              ^bb0(%arg11: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg12: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>):
                %35 = tag 1 %arg12 : tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>> -> <tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>>
                output %35 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>>
              }
              output %34 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>>
            }
            output %33 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>>
          }
          switch %7 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>> ^bb3, ^bb4
        ^bb3(%8: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %9: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %10: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb2
          %11:5 = dfg input extensions [] %8, %9, %10 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg7: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg8: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            const @const_41 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 0 : i6
            %29 = load_constant @const_41 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %30 = make_tuple()
            %31 = tag 0 %30 : tuple<> -> <tuple<>>
            output %31, %arg6, %29, %arg7, %arg8 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %11#0, %11#1, %11#2, %11#3, %11#4 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb5
        ^bb4(%12: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb2
          %13:2 = dfg input extensions [] %12 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            %29 = make_tuple()
            %30 = tag 0 %29 : tuple<> -> <tuple<>>
            output %30, %arg6 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %13#0, %13#1 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb1
        ^bb5(%14: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %15: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %16: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %17: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // 2 preds: ^bb3, ^bb6
          %18:5 = dfg input extensions [] %14, %15, %16, %17 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg7: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg8: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg9: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            const @const_49 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 0 : i6
            %29 = load_constant @const_49 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %30 = ext_op []"igt_s" %arg9, %29 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr.sum<tuple<>, tuple<>>
            output %30, %arg6, %arg7, %arg8, %arg9 : !hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %18#0, %18#1, %18#2, %18#3, %18#4 : !hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb6, ^bb7
        ^bb6(%19: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %20: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %21: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %22: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb5
          %23:5 = dfg input extensions [] %19, %20, %21, %22 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg7: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg8: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg9: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            %29 = ext_op []"iadd" %arg7, %arg8 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            const @const_54 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 1 : i6
            %30 = load_constant @const_54 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %31 = ext_op []"isub" %arg9, %30 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %32 = make_tuple()
            %33 = tag 0 %32 : tuple<> -> <tuple<>>
            output %33, %arg6, %29, %arg8, %31 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %23#0, %23#1, %23#2, %23#3, %23#4 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb5
        ^bb7(%24: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %25: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %26: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %27: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb5
          %28:4 = dfg input extensions [] %24, %25, %26, %27 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg7: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg8: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, %arg9: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            %29 = ext_op []"iadd" %arg6, %arg7 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            const @const_61 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 1 : i6
            %30 = load_constant @const_61 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %31 = ext_op []"isub" %arg8, %30 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %32 = make_tuple()
            %33 = tag 0 %32 : tuple<> -> <tuple<>>
            output %33, %29, %31, %arg9 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %28#0, %28#1, %28#2, %28#3 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb2
        }
        output %1 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
      }
      output %0 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
    }
  }
}

